### PR TITLE
JSDK-2442 update publish priority

### PR DESCRIPTION
**Pull request by makarandp0**
Re-opening #733 with integration tests added.

This PR contains the changes specified below. Once the `track_priority` MSP channel is available on VMS, I will add integration tests.

## Changes

- [x] Implementing `LocalTrackPublication.setPriority()`
- [x] Implementing `RemoteTrackPublication#publishPriorityChanged`
- [x] Unit tests
- [x] Integration tests
- [x] Update CHANGELOG.md

**Contributing to Twilio**

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

File: lib/signaling/v2/localparticipant.js [hunk 1]
```diff
@@ -50,6 +50,10 @@ class LocalParticipantV2 extends LocalParticipantSignaling {
         writable: true,
         value: 1
       },
+      _trackPrioritySignaling: {
```
Comment (manjeshbhargav) [inline lib/signaling/v2/localparticipant.js]: @makarandp0 Isn't this already present in `lib/signaling/v2/participant.js`?

File: lib/signaling/v2/localparticipant.js [hunk 2]
```diff
@@ -100,6 +104,16 @@ class LocalParticipantV2 extends LocalParticipantSignaling {
     return this;
   }
 
+  /**
+   * Set the {@link TrackPrioritySignaling}.
+   * @param {TrackPrioritySignaling} trackPrioritySignaling
+   * @returns {this}
+   */
+  setTrackPrioritySignaling(trackPrioritySignaling) {
```
Comment (manjeshbhargav) [inline lib/signaling/v2/localparticipant.js]: @makarandp0 Isn't this already present in `lib/signaling/v2/participant.js`?

File: lib/signaling/v2/localparticipant.js [hunk 3]
```diff
@@ -181,6 +203,15 @@ class LocalParticipantV2 extends LocalParticipantSignaling {
     }
   }
 
+  _updateTrackPriority(publication) {
```
Comment (manjeshbhargav) [inline lib/signaling/v2/localparticipant.js]: @makarandp0 Isn't this already present in `lib/signaling/v2/participant.js`?

File: lib/signaling/v2/room.js [hunk 4]
```diff
@@ -407,6 +407,18 @@ class RoomV2 extends RoomSignaling {
     return this;
   }
 
+  _setupTrackPriorityUpdates(trackPrioritySignaling) {
+    this._trackPrioritySignaling = trackPrioritySignaling;
+    trackPrioritySignaling.on('updated', (trackSid, publishOrSubscribe, priority) => {
+      const trackSignaling = this._getTrackSidsToTrackSignalings().get(trackSid);
+      if (trackSignaling) {
```
Comment (manjeshbhargav) [inline lib/signaling/v2/room.js]:
```suggestion
      if (trackSignaling && publishOrSubscribe === 'publish') {
```

File: test/integration/spec/localtracks.js [hunk 5]
```diff
@@ -75,3 +89,119 @@ const createLocalTracks = require('../../../lib/createlocaltrack');
     });
   });
 });
+
+describe('LocalParticipant', function() {
```
Comment (manjeshbhargav) [inline test/integration/spec/localtracks.js]:
```suggestion
describe('LocalTrackPublication', function() {
```

Comment (manjeshbhargav) [reply test/integration/spec/localtracks.js]: @makarandp0 This test should move to `test/integration/spec/localtrackpublication.js`.

**Comment by makarandp0** [comment]
Thanks @manjeshbhargav, for noticing the refactoring need. I have incorporated your feedbck now..

File: lib/signaling/participant.js [hunk 6]
```diff
@@ -163,13 +163,14 @@ class ParticipantSignaling extends StateMachine {
...
    * @param {Track.SID} trackSid
    * @param {?Track.Priority} priority
+   * @param {'publish'|'subscribe'} publishOrSubscribe
    * @returns {void}
    */
-  updateTrackPriority(trackSid, priority) {
+  updateTrackPriority(trackSid, priority, publishOrSubscribe) {
     if (this._trackPrioritySignaling) {
-      this._trackPrioritySignaling.sendTrackPriorityUpdate(trackSid, 'subscribe', priority);
+      this._trackPrioritySignaling.sendTrackPriorityUpdate(trackSid, publishOrSubscribe, priority);
```
Comment (manjeshbhargav) [inline lib/signaling/participant.js]: @makarandp0 I think whether to `publish` or `subscribe` can be initialized in the constructors of LocalParticipantV2 and RemoteParticipantV2:
```js
class LocalParticipantV2 {
  constructor() {
    Object.defineProperties(this, {
      _trackAction: {
        value: 'publish'
      }
    });
  }
}

class RemoteParticipantV2 {
  constructor() {
    Object.defineProperties(this, {
      _trackAction: {
        value: 'subscribe'
      }
    });
  }
}

class ParticipantSignaling {
  updateTrackPriority(trackSid, priority) {
    ...
    this._trackPrioritySignaling.sendTrackPriorityUpdate(trackSid, this._trackAction, priority);
    ...
    this._enqueuedPriorityUpdates.set(trackSid, priority);
  }
}
```
This way you don't have to store `publish` or `subscribe` in each enqueued entry.

File: test/integration/spec/localtrackpublication.js [hunk 7]
```diff
@@ -212,4 +220,110 @@ describe('LocalTrackPublication', function() {
...
+        createSyntheticAudioStreamTrack() || await createLocalAudioTrack({ fake: true }),
+        await createLocalVideoTrack(smallVideoConstraints),
+      ]));
+
+      [aliceLocal, bobLocal] = thoseRooms.map(room => room.localParticipant);
+      [aliceRemote, bobRemote] = [thisRoom.participants.get(aliceLocal.sid), thisRoom.participants.get(bobLocal.sid)];
+
+      // Alice publishes her tracks at low priority
+      // Bob publishes his tracks at standard priority
+      await Promise.all([
```
Comment (manjeshbhargav) [inline test/integration/spec/localtrackpublication.js]: @makarandp0 `waitFor`?

File: test/integration/spec/localtrackpublication.js [hunk 8]
```diff
@@ -212,4 +220,110 @@ describe('LocalTrackPublication', function() {
       });
     });
   });
+
+  describe('#setPriority', () => {
```
Comment (manjeshbhargav) [inline test/integration/spec/localtrackpublication.js]:
```suggestion
  (defaults.topology === 'peer-to-peer' ? describe.skip : describe)('#setPriority', () => {
```

File: test/integration/spec/localtrackpublication.js [hunk 9]
```diff
@@ -212,4 +220,110 @@ describe('LocalTrackPublication', function() {
...
+
+    afterEach(async () => {
+      [thisRoom, ...thoseRooms].forEach(room => room && room.disconnect());
+      [...aliceTracks, ...bobTracks].forEach(track => track.stop && track.stop());
+      if (thisRoom) {
+        await completeRoom(thisRoom.sid);
+      }
+    });
+
+    if (defaults.topology !== 'peer-to-peer') {
```
Comment (manjeshbhargav) [inline test/integration/spec/localtrackpublication.js]:
```suggestion
```

File: test/integration/spec/localtrackpublication.js [hunk 10]
```diff
@@ -212,4 +220,110 @@ describe('LocalTrackPublication', function() {
...
+
+        // expect Alice's track to get switched on, and Bob's track to get switched off
+        // also expect to receive publishPriorityChanged event on Bob's track.
+        await waitFor([
+          trackSwitchedOn(aliceRemoteVideoTrack),
+          trackSwitchedOff(bobRemoteVideoTrack),
+          trackPublishPriorityChanged(bobRemoteVideoTrackPublication)
+        ], 'Alice track to get switched On, and Bob Switched Off');
+      });
+    }
```
Comment (manjeshbhargav) [inline test/integration/spec/localtrackpublication.js]:
```suggestion
```

File: test/integration/spec/localtrackpublication.js [hunk 11]
```diff
@@ -212,4 +220,110 @@ describe('LocalTrackPublication', function() {
...
+
+    if (defaults.topology !== 'peer-to-peer') {
+      it('publisher can upgrade track\'s priority', async () => {
+        await waitFor([
+          trackSwitchedOn(bobRemoteVideoTrack),
+          trackSwitchedOff(aliceRemoteVideoTrack)
+        ], 'Bobs track to get switched On, and Alice Switched Off');
+
+        // Alice changes her track priority to high
+        aliceLocalVideoTrackPublication.setPriority(PRIORITY_HIGH);
```
Comment (manjeshbhargav) [inline test/integration/spec/localtrackpublication.js]:
```suggestion
        aliceLocalVideoTrackPublication.setPriority(PRIORITY_HIGH);
        assert.equal(aliceLocalVideoTrackPublication.priority, PRIORITY_HIGH);
```

File: test/integration/spec/localtrackpublication.js [hunk 12]
```diff
@@ -212,4 +220,110 @@ describe('LocalTrackPublication', function() {
...
+      });
+
+      it('publisher can downgrade track\'s priority', async () => {
+        await waitFor([
+          trackSwitchedOn(bobRemoteVideoTrack),
+          trackSwitchedOff(aliceRemoteVideoTrack)
+        ], 'Bobs track to get switched On, and Alice Switched Off');
+
+        // Bob changes his track priority to low
+        bobLocalVideoTrackPublication.setPriority(PRIORITY_LOW);
```
Comment (manjeshbhargav) [inline test/integration/spec/localtrackpublication.js]:
```suggestion
        bobLocalVideoTrackPublication.setPriority(PRIORITY_LOW);
        assert.equal(bobLocalVideoTrackPublication.priority, PRIORITY_LOW);
```

File: test/integration/spec/localtrackpublication.js [hunk 13]
```diff
@@ -212,4 +220,110 @@ describe('LocalTrackPublication', function() {
...
+
+        // Alice changes her track priority to high
+        aliceLocalVideoTrackPublication.setPriority(PRIORITY_HIGH);
+
+        // expect Alice's track to get switched on, and Bob's track to get switched off
+        await waitFor([
+          trackSwitchedOn(aliceRemoteVideoTrack),
+          trackSwitchedOff(bobRemoteVideoTrack),
+          trackPublishPriorityChanged(aliceRemoteVideoTrackPublication)
+        ], 'Alice track to get switched On, and Bob Switched Off');
```
Comment (manjeshbhargav) [inline test/integration/spec/localtrackpublication.js]:
```suggestion
        ], 'Alice track to get switched On, and Bob Switched Off');
        
        assert.equal(aliceRemoteVideoTrackPublication.publishPriority, PRIORITY_HIGH);
```

File: test/integration/spec/localtrackpublication.js [hunk 14]
```diff
@@ -212,4 +220,110 @@ describe('LocalTrackPublication', function() {
...
+        // Bob changes his track priority to low
+        bobLocalVideoTrackPublication.setPriority(PRIORITY_LOW);
+
+        // expect Alice's track to get switched on, and Bob's track to get switched off
+        // also expect to receive publishPriorityChanged event on Bob's track.
+        await waitFor([
+          trackSwitchedOn(aliceRemoteVideoTrack),
+          trackSwitchedOff(bobRemoteVideoTrack),
+          trackPublishPriorityChanged(bobRemoteVideoTrackPublication)
+        ], 'Alice track to get switched On, and Bob Switched Off');
```
Comment (manjeshbhargav) [inline test/integration/spec/localtrackpublication.js]:
```suggestion
        ], 'Alice track to get switched On, and Bob Switched Off');
        
        assert.equal(bobRemoteVideoTrackPublication.publishPriority, PRIORITY_LOW);
```

File: test/integration/spec/remotetracks.js [hunk 15]
```diff
@@ -108,7 +108,7 @@ describe('RemoteParticipant', function() {
 
     // eslint-disable-next-line no-warning-comments
     // TODO: enable these tests when track_priority MSP is available in prod
-    if (defaults.topology !== 'peer-to-peer' && defaults.environment === 'dev') {
+    if (defaults.topology !== 'peer-to-peer' && defaults.environment === 'stage') {
```
Comment (manjeshbhargav) [inline test/integration/spec/remotetracks.js]: @makarandp0 This condition should be applied at the top-level `describe`:
```js
(defaults.topology === 'peer-to-peer' || defaults.environment === 'prod'
  ? describe.skip
  : describe)('#setPriority', () => {...});
```
Also, please change `describe('RemoteParticipant', ...)` to `describe('RemoteVideoTrack', ...)`.

File: test/integration/spec/remotetracks.js [hunk 16]
```diff
@@ -22,11 +22,16 @@ function getTracksOfKind(participant, kind) {
...
-describe('RemoteParticipant', function() {
+
+describe('RemoteVideoTrack', function() {
   // eslint-disable-next-line no-invalid-this
   this.timeout(60000);
 
-  describe('#setPriority', () => {
+  // eslint-disable-next-line no-warning-comments
+  // TODO: enable these tests when track_priority MSP is available in prod
+  (defaults.topology === 'peer-to-peer' || defaults.environment === 'prod'
```
Comment (manjeshbhargav) [inline test/integration/spec/remotetracks.js]: @makarandp0 On second thought, I think the following tests should run for all values of topology and environment:
```
- can setpriority null on audio track
- can setpriority high on audio track
- can setpriority low on audio track
- can setpriority standard on audio track
- can setpriority null on video track
- can setpriority high on video track
- can setpriority low on video track
- can setpriority standard on video track
- can setpriority null on data track
- can setpriority high on data track
- can setpriority low on data track
- can setpriority standard on data track
- throws for an invalid priority value: undefined
- throws for an invalid priority value: foo
- throws for an invalid priority value: 42
```
The other tests should be conditionally enabled.

**Pull request merged**